cephfs: implement snapdiff

ifed01 opened the pull request:
This patch supersedes both #42517 and #43328 and implements API to access snapshot delta (aka snapdiff) via
new libcephfs API method: readdir_snapdiff.

New test cases "LibCephFS.SnapDiff*" can serve as an example of new API calls.

Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>
Signed-off-by: Denis Barahtanov <denis.barahtanov@croit.io>

<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

vshankar commented:
@ifed01 I most likely have some unpublished comments in the other pr by @denisb-croit.

I'll go through this anyway.

ifed01 commented:
@vshankar - gentle reminder about the PR ;)

vshankar commented:
> @vshankar - gentle reminder about the PR ;)

ACK.

vshankar commented:
jenkins test make check arm64

vshankar commented:
jenkins test make check arm64

vshankar commented:
@ifed01 - sorry for the delay. I have started reviewing and testing these changes starting last Friday - will let you know how things go.

BTW, I'm planning to include these changes once quincy is branched off. I hope that is ok?

ifed01 commented:
> @ifed01 - sorry for the delay. I have started reviewing and testing these changes starting last Friday - will let you know how things go.
No worries, thanks in advance.
> 
> BTW, I'm planning to include these changes once quincy is branched off. I hope that is ok?
Sure.

ifed01 commented:
@vshankar - any updates? ;)

vshankar commented on src/include/cephfs/libcephfs.h:
```diff
@@ -604,6 +604,23 @@ int ceph_readdir_r(struct ceph_mount_info *cmount, struct ceph_dir_result *dirp,
+ *          is empty.  This pointer should not be freed by the caller, and is only safe to
+ *          access between return and the next call to ceph_readdir or ceph_closedir.
+ */
+int ceph_readdir_snapdiff(struct ceph_mount_info* cmount,
```
It's looks a bit unusual to pass in the snap-id of the _other_ snapshot against which the delta has to be built. Could this be the snapshot name?

FWIW, this interface can accept two directory stream pointers (one for each snapshot) and a path for which the entries are to be built.

vshankar replied:
@ifed01 @batrick I'm open to suggestions on how the API would look like? Basically, I would like to avoid passing in snap-ids if possible.

ifed01 replied:
@vshankar - I've reworked the API to accept snapshot root(!) dir stream pointer instead of snapid.
There is still a snapid returned by the function to distinguish removed and existing entries. Probably we can get rid off it as well and replace with an explicit bool flag... Do you think we need should do that?

ifed01 replied:
@batrick - I refactored this part a bit, please take a look. Is that what you suggested?

vshankar reviewed (commented):
@ifed01 I have mostly reviewed changes in MDS in Denis's original PR which are carry-forwarded in this PR, so, just commenting on the interface for now...

(I have tried playing/testing the original PR which worked well)

ifed01 commented:
> @ifed01 I have mostly reviewed changes in MDS in Denis's original PR which are carry-forwarded in this PR, so, just commenting on the interface for now...
@vshankar - thanks for taking a look. Could you please elaborate what PR are you referring as the original one: #42517 (which has got just the "tilda" interface) or #43328 (which is a dual-API evolution). I can't see any comments from your side for the latter - is it that good or you just haven't publish the comments? ;)

vshankar commented:
> @vshankar - thanks for taking a look. Could you please elaborate what PR are you referring as the original one: #42517 (which has got just the "tilda" interface) or #43328 (which is a dual-API evolution). I can't see any comments from your side for the latter - is it that good or you just haven't publish the comments? ;)

I'm talking about Denis's PR https://github.com/ceph/ceph/pull/42517 where I had some comments that Denis took care of.

ifed01 commented:
jenkins test make check

ifed01 commented:
> > @vshankar - thanks for taking a look. Could you please elaborate what PR are you referring as the original one: #42517 (which has got just the "tilda" interface) or #43328 (which is a dual-API evolution). I can't see any comments from your side for the latter - is it that good or you just haven't publish the comments? ;)
> 
> I'm talking about Denis's PR #42517 where I had some comments that Denis took care of.

It looks like those comments have been answered/resolved? Do you see anything still open for discussion there ?

vshankar commented:
> It looks like those comments have been answered/resolved? Do you see anything still open for discussion there ?

Nothing. I'm waiting for quincy to branch out as mentioned in https://github.com/ceph/ceph/pull/43546#issuecomment-986572349.

Also, I've requested @batrick for review.

ifed01 commented:
jenkins test api

ifed01 commented:
jenkins test docs

ifed01 commented:
@batrick - gentle reminder.

batrick commented on src/mds/Mutation.h:
```diff
@@ -389,6 +389,30 @@ struct MDRequestImpl : public MutationImpl {
+  // Primary snapshot id is encoded as a lower 32-bits of the effective id.
+  // And higher 32-bits keep the secondary (either first or last snapid for
+  // updated or removed entries respectively).
+  // The following function properly builds effective snapid as per above.
```
This needs some examples perhaps and more explanation. I confess I'm mystified by this comment.

ifed01 replied:
>I think better would be to create an API for building a snapdiff stream. Something like:

>What do you think?
I refactored this part a bit, please take a look.

batrick commented on src/mds/Server.cc:
```diff
@@ -236,6 +236,8 @@ void Server::create_logger()
                    "Request type remove snapshot latency");
   plb.add_time_avg(l_mdss_req_renamesnap_latency, "req_renamesnap_latency",
                    "Request type rename snapshot latency");
+  plb.add_time_avg(l_mdss_req_snapdiff_latency, "req_snapdiff_latency",
```
mode changed on this file too

batrick reviewed (changes requested):
I won't be able to fully go through the Server.cc/Client.cc bits now as I'm about to go on extended leave. So, I'm going to give some broad requests to help get this merged:

* Please add dev docs/comments. It needs to be crystal clear what the request is at each layer (userspace, libcephfs, Client, RPC, Server). Snapshots are one of the most poorly understood parts of the MDS; we really need to build confidence in what we're merging here makes sense.
* User docs are missing. Is there a tool by SUSE or Croit planning to use this? Has it already been written? Can cephfs-mirror be enhanced with this? Is it expected that a backup tool would only use libcephfs/Client to do snapdiff or could a shell script also work?
* The tests look rigorous but are difficult to follow. There needs to be exposition about what each test is doing. Some tests could probably be broken up into smaller tests. Keep in mind that tests are one of the first things a developer will end up looking at when things go wrong. These tests are hard to pickup for someone tracking down a bug.
* A shellscript test would be good for exercising userspace mounts.
* Should there not be a python libcephfs binding?
* Do we need to piggy back on the .snap directory for snapdiff?
* In the results, how does the user distinguish between a changed file beginning with `~` and a file name that happens to begin with `~`? Is it not unnecessary to include unchanged dentries?
* What is the feature bit explicitly protecting? To me it looks like there is a new RPC. Are you avoiding the MDS returning results an older client does not understand?

With all that said, snapdiff is something we've wanted to add and I'm excited to see this PR. Thanks for your work on it!

lukasbruna commented:
Hello,
after trying out the diff on a directory of approximately 3,2 million (empty) files, I found out the snap diff feature is both time and memory dependent on the amount of files in the directory, not on the amount of files changed between the 2 snapshots.
Is that the architecture of this feature or am I doing something wrong?

- diff on 3,2 million file snapshots takes 5 minutes to list difference of 1 file
![image](https://user-images.githubusercontent.com/88100740/159482619-5c016806-5dae-43bd-80d4-ce38c86c6bf4.png)
(diff on 100k file snapshots in the same cluster takes 0.083s)

- 4G of RAM - diff on 2 100k files snapshots - no files changed, **the memory is not freed after the diff is done**
![memory_dependency](https://user-images.githubusercontent.com/88100740/159482059-37cb2941-5674-4afa-9661-6a252cd6c9b6.png)

vshankar commented:
@ifed01 Mind taking a look at @batrick comment when you get time? I understand that this PR has been going back-and-forth for sometime, however, given the importance and usefulness of the functionality that's being implemented, the interface and the implementation does require some hardening to avoid rework in the future. If you could address the comments and push an update, I can prioritize further reviews and running in through our tests.

ifed01 commented:
@vshankar - yeah, will come back to this patch in a couple of days

vshankar commented:
> @vshankar - yeah, will come back to this patch in a couple of days

Thx! Just FYI, I'm on PTO next week. Feel free to ping @ceph/cephfs for review requests.

ifed01 commented:
> > @vshankar - yeah, will come back to this patch in a couple of days
> 
> Thx! Just FYI, I'm on PTO next week. Feel free to ping @ceph/cephfs for review requests.
OK, but I think it will take next week to resolve the comments hence you're back on time ;)

gregsfortytwo commented:
We also need to figure out the memory usage described: https://github.com/ceph/ceph/pull/43546#issuecomment-1075120410

Is it just that the client grabbed a bunch of caps and isn't ditching the snapshot caps because the MDS doesn't ever need them back? Or is there some kind of memory leak we're missing?

ifed01 commented:
>Please add dev docs/comments. It needs to be crystal clear what the request is at each layer (userspace, libcephfs, Client, RPC, Server). Snapshots are one of the most poorly understood parts of the MDS; we really need to build confidence in what we're merging here makes sense.
I'm still on that.

>User docs are missing. Is there a tool by SUSE or Croit planning to use this? Has it already been written? Can cephfs-mirror be enhanced with this? Is it expected that a backup tool would only use libcephfs/Client to do snapdiff or could a shell script also work?
I'd prefer to add user docs while we're done with the implementation to avoid redundant work on maintaining actual docs during dev cycles.
No client tools are available atm. But we're planning to use that at Croit indeed.
Given two snapdiff APIs are provided I expect both shell scripts and specific tools can use it. Likely cephfs-mirror to benefit from that too.

>The tests look rigorous but are difficult to follow. There needs to be exposition about what each test is doing. Some tests could probably be broken up into smaller tests. Keep in mind that tests are one of the first things a developer will end up looking at when things go wrong. These tests are hard to pickup for someone tracking down a bug.

Done.


>A shellscript test would be good for exercising userspace mounts.
Will do. Are you aware about any existing cephfs shell script tests to use as a basis?

>Should there not be a python libcephfs binding?

Please elaborate? Do you mean extending python libcephfs binding with new funciton(s)?

>Do we need to piggy back on the .snap directory for snapdiff?

That's definitely not mandatory but I doubt having another fake ".snapdiff" (?) folder would make things better...

>In the results, how does the user distinguish between a changed file beginning with ~ and a file name that happens to begin with ~? 
Not implemented atm, thinking about escaping starting tildas by duplicating each of them

>Is it not unnecessary to include unchanged dentries?
You mean including unchanged files? What for?

>What is the feature bit explicitly protecting? To me it looks like there is a new RPC. Are you avoiding the MDS returning results an older client does not understand?

That's rather a sort of protection for newer clients to avoid handling readdir results from legacy MDS-s in a new (snapdiff) manner: primarily stripping off  that starting tilda for removed snapdiff entries. With the above mentioned tilda escaping introduction we might additionally need this flag.
But may be a bit too redundant indeed...

ifed01 commented:
@vshankar @batrick - I rebased the patch and resolved some comment. With a few still on my plate. Will proceed working on that  once back from vacation in mid-June. Would be nice to get some answers and brief intermediate review on the changes by then ;)

ifed01 commented:
> After trying out the diff on a directory of approximately 3,2 million (empty) files, I found out the snap diff feature is both time and memory dependent on the amount of files in the directory, not on the amount of files changed between the 2 snapshots.
Is that the architecture of this feature or am I doing something wrong?

@lukasbruna - I'm very sorry for the late response - totally missed your comment. 
For now this feature's computational complexity depends on the amount of entries in the folder indeed. As for RAM consumption it's rather depends on the existing directory caching capabilities so I'd bet on some misbehavior/bug with that stuff. But surely this can be due to the patch as well... If you still have that setup - would you check RAM consumption if this directory content is listed in a regular manner with no snapdiff involved. Would it expose the similar behavior?

batrick commented on src/test/libcephfs/snapdiff.cc:
```diff
@@ -0,0 +1,1831 @@
+void prepareSnapDiffCases(TestMount& test_mount)
+{
+  //************ snap1 *************
+  ceph_assert(0 < test_mount.write_full("a", "file 'a' v1"));
```
These should be using `ASSERT_EQ`. I see this is a static procedure. You could make this a method in a parent test class.

batrick reviewed (changes requested):
(Fixing the formatting, I think:)

> > Please add dev docs/comments. It needs to be crystal clear what the request is at each layer (userspace, libcephfs, Client, RPC, Server). Snapshots are one of the most poorly understood parts of the MDS; we really need to build confidence in what we're merging here makes sense.
> I'm still on that.
> 
> > User docs are missing. Is there a tool by SUSE or Croit planning to use this? Has it already been written? Can cephfs-mirror be enhanced with this? Is it expected that a backup tool would only use libcephfs/Client to do snapdiff or could a shell script also work?
> I'd prefer to add user docs while we're done with the implementation to avoid redundant work on maintaining actual docs during dev cycles.
>
> No client tools are available atm. But we're planning to use that at Croit indeed.
> Given two snapdiff APIs are provided I expect both shell scripts and specific tools can use it. Likely cephfs-mirror to benefit from that too.
> 
> > The tests look rigorous but are difficult to follow. There needs to be exposition about what each test is doing. Some tests could probably be broken up into smaller tests. Keep in mind that tests are one of the first things a developer will end up looking at when things go wrong. These tests are hard to pickup for someone tracking down a bug.
> 
> Done.

Will review this.

> > A shellscript test would be good for exercising userspace mounts.
> Will do. Are you aware about any existing cephfs shell script tests to use as a basis?

See `qa/workunits/fs/`. Various yamls refer to these in `qa/suites/fs`.

> > Should there not be a python libcephfs binding?
> 
> Please elaborate? Do you mean extending python libcephfs binding with new funciton(s)?

Yes

> > Do we need to piggy back on the .snap directory for snapdiff?
> 
> That's definitely not mandatory but I doubt having another fake ".snapdiff" (?) folder would make things better...

Even though the format you've chosen is obscure (e.g. `.snap/.~diff=snap1.~diff=snap3`), it's generally better to build a separate API/interface than to re-use. I think we'll have more of these alternative directories for future CephFS-specific operations so it would not be weird.

In any case, I don't feel strongly about this either way.

> > In the results, how does the user distinguish between a changed file beginning with ~ and a file name that happens to begin with ~?
> > Not implemented atm, thinking about escaping starting tildas by duplicating each of them

That's not viable. I think you need a starting char for each file name. Either `-` or `~`. Or whatever.

> > Is it not unnecessary to include unchanged dentries?
> You mean including unchanged files? What for?

Nevermind, I misunderstood.
 
> > What is the feature bit explicitly protecting? To me it looks like there is a new RPC. Are you avoiding the MDS returning results an older client does not understand?
> 
> That's rather a sort of protection for newer clients to avoid handling readdir results from legacy MDS-s in a new (snapdiff) manner: primarily stripping off that starting tilda for removed snapdiff entries. With the above mentioned tilda escaping introduction we might additionally need this flag. But may be a bit too redundant indeed...

I think you can just create a new client request op instead of piggybacking on readdir op. Then an old MDS correctly replies `ENOTSUPP`.

ifed01 commented:
@batrick, @vshankar - I updated the PR and removed the initial implementation through fake folder. New readdir_snapdiff API is supported for now only. Python binding on top of it has been added as well. Commits squashed and polished. Please take a look and review. 
Additionally I'm planning to provide some docs once we're fine with the implementation

ifed01 commented:
> We also need to figure out the memory usage described: [#43546 (comment)](https://github.com/ceph/ceph/pull/43546#issuecomment-1075120410)
> 
> Is it just that the client grabbed a bunch of caps and isn't ditching the snapshot caps because the MDS doesn't ever need them back? Or is there some kind of memory leak we're missing?

Well, I've performed a bunch of tests against an updated(not the one @lukasbruna tried) snapdiff implementation. From time to time (with no systematic pattern) I can see transient MDS (not client) memory consumption growth which goes away once the diff collecting process is running. MDS's cache drop/status commands (as well as dump_mempool) clearly indicate that the consumption is bound to MDS cache. Additionally from my observations RAM consumption is not increased over every new delta collection  So I presume this is a valid behavior caused by MDS cache growth and it might be kept as-is unless more severe concerns/complaints are provided

GittyBaer commented:
What is the status of this commit?

We could really use this in our environment :-)

ifed01 commented:
> What is the status of this commit?
> 
> We could really use this in our environment :-)

Pending @batrick  review... Patrick, sorry I'm using this as a chance to ping you once again ;)

vshankar commented:
> What is the status of this commit?

Its pending review after @ifed01 latest update. I'd say its really getting closer to being done.

vshankar commented:
jenkins retest this please

ifed01 commented on src/mds/Server.cc:
```diff
@@ -11157,3 +11334,265 @@ void Server::dump_reconnect_status(Formatter *f) const
+	  << dn->first << "/" << dn->last << dendl;
+	exists = false;
+      }
+      bool r = add_result_cb(dn, in, exists);
```
That's true, looks like there is no way to detect if something has changed within a directory by inspecting exclusively the dir entry itself . Hence these deeper dives is left to the caller

ifed01 commented on src/test/libcephfs/snapdiff.cc:
```diff
@@ -0,0 +1,1471 @@
+    expected.emplace_back("dirA", snapid2);
+    expected.emplace_back("dirB", snapid2);
+    expected.emplace_back("dirC", snapid1);
+    expected.emplace_back("dirD", snapid2);
```
True, see my comment below

ifed01 commented:
jenkins test make check

ifed01 commented:
jenkins test api

ifed01 commented on src/mds/Server.cc:
```diff
@@ -11142,6 +11166,159 @@ void Server::_renamesnap_finish(MDRequestRef& mdr, CInode *diri, snapid_t snapid
+      }
+    } else {
+      offset_str.clear();
+      newfg = diri->dirfragtree[fg.value()];
```
Here is a log snippet for the case:
2023-01-20T19:36:48.984+0300 7f289c42b6c0  0 mds.0.server  frag * offset '' offset_hash 0 flags 1
2023-01-20T19:36:48.984+0300 7f289c42b6c0  0 mds.0.server  adjust frag * -> 000* fragtree_t(*^3)

ifed01 commented on src/mds/Server.cc:
```diff
@@ -11142,6 +11166,159 @@ void Server::_renamesnap_finish(MDRequestRef& mdr, CInode *diri, snapid_t snapid
+    << " offset_hash " << offset_hash << " flags " << req_flags << dendl;
+
+  // does the frag exist?
+  if (diri->dirfragtree[fg.value()] != fg) {
```
Leaving this question to @batrick

ifed01 commented:
@vshankar - hopefully I resolved most of you last comments, mind taking another look?

vshankar commented:
> @vshankar - hopefully I resolved most of you last comments, mind taking another look? 

I'll have a look.

ifed01 commented:
> > @vshankar - hopefully I resolved most of you last comments, mind taking another look?
> 
> I'll have a look.

@vshankar - gentle reminder ;)

vshankar commented:
> > > @vshankar - hopefully I resolved most of you last comments, mind taking another look?
> > 
> > 
> > I'll have a look.
> 
> @vshankar - gentle reminder ;)

On my list for testing now that tests can be resumed in teuthology.

vshankar commented:
jenkins retest this please

vshankar commented on src/mds/Server.cc:
```diff
@@ -11250,6 +11274,159 @@ void Server::_renamesnap_finish(MDRequestRef& mdr, CInode *diri, snapid_t snapid
+  unsigned max_bytes = req->head.args.snapdiff.max_bytes;
+  if (!max_bytes)
+    // make sure at least one item can be encoded
+    max_bytes = (512 << 10) + g_conf()->mds_max_xattr_pairs_size;
```
@ifed01 This config option has been removed with #46357 and is now a part of the mds map. You'd need to switch using `mds->mdsmap->get_max_xattr_size()`.

ifed01 replied:
@vshankar - fixed, thanks!

vshankar commented:
fs suite run: https://pulpito.ceph.com/?branch=wip-vshankar-testing-20230330.105356

(test run review is in progress - will update once done)

vshankar commented:
jenkins retest this please

vshankar commented:
@ifed01 https://jenkins.ceph.com/job/ceph-windows-pull-requests/21495/consoleFull#17042323442a811ea2-3e7b-466b-84b4-d13df7e35809

FWIW, this builds fine in shaman.

ifed01 commented:
jenkins test docs
jenkins render docs

ifed01 commented:
jenkins test docs

petrutlucian94 commented on src/test/libcephfs/snapdiff.cc:
```diff
@@ -0,0 +1,1669 @@
+ */
+
+#include "gtest/gtest.h"
+#include "include/cephfs/libcephfs.h"
```
The Windows job fails with:

```
In file included from /home/ubuntu/ceph/src/test/libcephfs/snapdiff.cc:15:
/home/ubuntu/ceph/src/include/cephfs/libcephfs.h:161:29: error: ‘uid_t’ was not declared in this scope; did you mean ‘pid_t’?
  161 | UserPerm *ceph_userperm_new(uid_t uid, gid_t gid, int ngids, gid_t *gidlist);
      |                             ^~~~~
      |                             pid_t
```

The reason is that uid_t is not available on Windows, being defined by ``src/include/compat.h``. Unfortunately we can't include ``compat.h`` in ``libcephfs.h``, so we can either include ``compat.h`` here or move the ``uid_t`` and ``gid_t`` definitions to ``src/include/win32/fs_compat.h``.

petrutlucian94 commented on src/test/libcephfs/snapdiff.cc:
```diff
@@ -0,0 +1,1669 @@
+      for_each_readdir(relpath0,
+        [&] (const dirent* dire) {
+          string relpath = concat_path(relpath0, dire->d_name);
+          if (dire->d_type == DT_REG) {
```
The ``dirent`` structure does not include ``d_type`` on Windows. Can we use ``ceph_readdirplus_r`` instead?

Here's a diff:

```
diff --git a/src/include/compat.h b/src/include/compat.h
index 517e9d4fcd5..b19662e4d6d 100644
--- a/src/include/compat.h
+++ b/src/include/compat.h
@@ -251,9 +251,6 @@ typedef unsigned int uint;

 typedef _sigset_t sigset_t;

-typedef unsigned int uid_t;
-typedef unsigned int gid_t;
-
 typedef unsigned int blksize_t;
 typedef unsigned __int64 blkcnt_t;
 typedef unsigned short nlink_t;
diff --git a/src/include/win32/fs_compat.h b/src/include/win32/fs_compat.h
index 2fce1b72e85..635ae8bb6ab 100644
--- a/src/include/win32/fs_compat.h
+++ b/src/include/win32/fs_compat.h
@@ -40,3 +40,6 @@

 #define XATTR_CREATE  1
 #define XATTR_REPLACE 2
+
+typedef unsigned int uid_t;
+typedef unsigned int gid_t;
diff --git a/src/test/libcephfs/snapdiff.cc b/src/test/libcephfs/snapdiff.cc
index 72e7d77ec04..ec87abc0f68 100644
--- a/src/test/libcephfs/snapdiff.cc
+++ b/src/test/libcephfs/snapdiff.cc
@@ -148,7 +148,7 @@ public:
   }

   int for_each_readdir(const char* relpath,
-    std::function<bool(const dirent* dire)> fn)
+    std::function<bool(const dirent*, const struct ceph_statx*)> fn)
   {
     auto subdir_path = make_file_path(relpath);
     struct ceph_dir_result* ls_dir;
@@ -156,17 +156,34 @@ public:
     if (r != 0) {
       return r;
     }
-    struct dirent* result;
-    while( nullptr != (result = ceph_readdir(cmount, ls_dir))) {
-      if (strcmp(result->d_name, ".") == 0 ||
-          strcmp(result->d_name, "..") == 0) {
+
+    while (1) {
+      struct dirent result;
+      struct ceph_statx stx;
+
+      r = ceph_readdirplus_r(
+        cmount, ls_dir, &result, &stx, CEPH_STATX_BASIC_STATS,
+        0,
+        NULL);
+      if (!r)
+        break;
+      if (r < 0) {
+        std::cerr << "ceph_readdirplus_r failed, error: "
+                  << r << std::endl;
+        return r;
+      }
+
+      if (strcmp(result.d_name, ".") == 0 ||
+          strcmp(result.d_name, "..") == 0) {
         continue;
       }
-      if (!fn(result)) {
+      if (!fn(&result, &stx)) {
         r = -EINTR;
         break;
       }
     }
     ceph_assert(0 == ceph_closedir(cmount, ls_dir));
     return r;
   }
@@ -176,7 +193,7 @@ public:
     vector<string> expected(expected0);
     auto end = expected.end();
     int r = for_each_readdir(relpath,
-      [&](const dirent* dire) {
+      [&](const dirent* dire, const struct ceph_statx* stx) {

         std::string name(dire->d_name);
         auto it = std::find(expected.begin(), end, name);
@@ -281,13 +298,13 @@ public:
   {
     int r =
       for_each_readdir(relpath0,
-        [&] (const dirent* dire) {
+        [&] (const dirent* dire, const struct ceph_statx* stx) {
           string relpath = concat_path(relpath0, dire->d_name);
-          if (dire->d_type == DT_REG) {
-            unlink(relpath.c_str());
-          } else if (dire->d_type == DT_DIR) {
+          if (S_ISDIR(stx->stx_mode)) {
             purge_dir(relpath.c_str());
             rmdir(relpath.c_str());
+          } else {
+            unlink(relpath.c_str());
           }
           return true;
         });
```

vshankar replied:
Thanks @petrutlucian94 for noticing the windows build failure. I _really_ need to start paying more attention to those (at least the builds).

ifed01 replied:
Indeed, thanks a lot @petrutlucian94 for the diffs and suggestions. Now it looks fine for Win build as well!

vshankar commented:
> fs suite run: https://pulpito.ceph.com/?branch=wip-vshankar-testing-20230330.105356
> 
> (test run review is in progress - will update once done)

Failures are unrealted. Looking at 3 dead runs now.

BTW, the docs jenkins test are failing - those look unrelated too.

vshankar commented:
failed jobs rerun: https://pulpito.ceph.com/vshankar-2023-04-06_04:14:11-fs-wip-vshankar-testing-20230330.105356-testing-default-smithi/

vshankar reviewed (approved):
https://tracker.ceph.com/projects/cephfs/wiki/Main#10-Apr-2023

vshankar commented:
jenkins retest this please

vshankar commented:
@ifed01 mind rebasing this please? The docs build are failing...

ifed01 commented:
jenkins test make check

ifed01 commented:
jenkins test ceph windows tests

ifed01 commented:
jenkins test ceph windows

ifed01 commented:
jenkins test  windows

vshankar commented:
jenkins retest this please

ifed01 commented:
jenkins test windows

petrutlucian94 commented:
Thanks for issuing a Windows recheck. For what is worth, the Windows job is currently failing due to an unrelated regression: https://github.com/ceph/ceph/pull/51043

ifed01 commented:
> Thanks for issuing a Windows recheck. For what is worth, the Windows job is currently failing due to an unrelated regression: #51043

OK, good to know. Thanks!

vshankar commented:
> Thanks for issuing a Windows recheck. For what is worth, the Windows job is currently failing due to an unrelated regression: #51043

Thx @petrutlucian94. Since the failure is unrelated, this is good to merge then :)

ifed01 commented:
> > Thanks for issuing a Windows recheck. For what is worth, the Windows job is currently failing due to an unrelated regression: #51043
> 
> Thx @petrutlucian94. Since the failure is unrelated, this is good to merge then :)

OK, I'm going ahead and merge..
